### PR TITLE
Add code comment for StuckSplitTasksInterrupter race condition

### PR DESCRIPTION
**Pull request by leetcode-1533**
## Description

Add code comment for StuckSplitTasksInterrupter race condition.
More details at https://github.com/trinodb/trino/pull/13272

## Additional context and related issues
https://github.com/trinodb/trino/pull/13272

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java [hunk 1]
```diff
@@ -705,6 +706,29 @@ private Optional<StuckSplitTasksInterrupter> createStuckSplitTasksInterrupter(
                         taskExecutor));
     }
 
+    /**
+     * The class detect and interrupt runaway splits threads. It interrupts threads via failing the task that holding the split
+     * and relying on {@link PrioritizedSplitRunner#destroy()} method to actually interrupt the responsible thread.
+     * It detects at the frequency of {@link StuckSplitTasksInterrupter#stuckSplitsDetectionInterval}.
```
Comment (phd3) [inline core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java:712]: Can we swap order of the statement with the next one? and minor reword suggestion of  something like:

```
The detection is invoked periodically with the frequency of {@link StuckSplitTasksInterrupter#stuckSplitsDetectionInterval}.
```

File: core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java [hunk 2]
```diff
@@ -882,6 +886,13 @@ public Set<TaskId> getStuckSplitTaskIds(Duration processingDurationThreshold, Pr
                 .filter(filter).map(RunningSplitInfo::getTaskId).collect(toImmutableSet());
     }
 
+    /**
+     * A class representing a split that is running on the taskRunner.
```
Comment (phd3) [inline core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:890]: nit: `TaskRunner` or `task runner`

File: core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java [hunk 3]
```diff
@@ -882,6 +886,13 @@ public Set<TaskId> getStuckSplitTaskIds(Duration processingDurationThreshold, Pr
...
+     * It has thread objects that get assigned while assigning the split
```
Comment (phd3) [inline core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:891]: It has a Thread object that gets assigned ...

File: core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java [hunk 4]
```diff
@@ -882,6 +886,13 @@ public Set<TaskId> getStuckSplitTaskIds(Duration processingDurationThreshold, Pr
...
+     * to the taskRunner. However, since the assigned thread is part of a thread pool that
```
Comment (phd3) [inline core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:892]: How about we  reword this as 

```
However, when the TaskRunner moves to a different split, the thread stored here will not remain assigned to this split anymore.  
```

File: core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java [hunk 5]
```diff
@@ -488,6 +488,10 @@ private synchronized void recordLeafSplitsSize()
         this.lastLeafSplitsSize = allSplits.size() - intermediateSplits.size();
     }
 
+    /**
+     * A class acting as a worker as part of threadpool that keep pulling pending splits,
```
Comment (phd3) [inline core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:492]: I think since it implements `Runnable`, the first part is kind of clear. May be just say something like below or remove

```
Keeps processing split one-by-one from the list of waiting splits.
```

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java [hunk 6]
```diff
@@ -705,6 +706,29 @@ private Optional<StuckSplitTasksInterrupter> createStuckSplitTasksInterrupter(
...
+     * A thread get interrupted once the split had been processed beyond {@link StuckSplitTasksInterrupter#interruptStuckSplitTasksTimeout} and the split threaddump matches with
+     * {@link StuckSplitTasksInterrupter#stuckSplitStackTracePredicate} <p>
+     * There is a potential race condition for this {@link StuckSplitTasksInterrupter} class.
+     * The problematic flow is that we may kill a task that is long-running, but not really stuck on JONI code. Consider the following example:
+     * <ol>
+     * <li>We find long-running splits; we get A, B, C.</li>
```
Comment (phd3) [inline core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java:718]: thanks for adding these details!

**Comment by phd3** [comment]
cc @arhimondr

**Comment by leetcode-1533** [comment]
updated thanks!

File: core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java (hunk 5 above)
Comment (leetcode-1533) [reply core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java]: removed.

**Review by phd3 — APPROVED** [review]
Made some minor language changes, will merge after CI goes through

**Comment by phd3** [comment]
hit #14814, before the fix was merged
